further_link/util/bluetooth/server.py

```diff
@@ -0,0 +1,166 @@
+        if self.server is None:
+            raise Exception("Bluetooth server not started")
+
+        if isinstance(value, str):
+            value = to_byte_array(value)
+
+        logging.debug(f"Writing '{value}' to characteristic {uuid}")
+        characteristic = self.server.get_characteristic(uuid)
+        characteristic.value = value
+
+        # Notify subscribers
+        self.server.update_value(self._get_service_uuid(uuid), uuid)
+
+    def read_value(self, uuid: str) -> Optional[bytearray]:
+        if self.server is None:
+            raise Exception("Bluetooth server not started")
+
+        value = bytearray()
+        service_uuid = self._get_service_uuid(uuid)
+        service = self.server.services.get(service_uuid)
+        if service:
+            char = service.get_characteristic(uuid)
+            value = self._read_request(char)
+        return value
+
+    def _read_request(self, characteristic: BlessGATTCharacteristic, **kwargs):
+        """method required by bless to handle read requests"""
+
+        logging.debug(f"Read request for characteristc {characteristic}")
+        value = decode_value(characteristic.value)
+
+        # callback on read requests will update the value of the characteristic before returning
+        callback = (
+            self.config.tree.get(self._get_service_uuid(characteristic.uuid), {})
+            .get(characteristic.uuid, {})
+            .get("ReadHandler")
+        )
+        if callback and callable(callback):
+            logging.debug(f"Executing ReadHandler callback for {characteristic.uuid}")
+            try:
+                value = str(callback())
+                # TODO: handle long messages using ChunkedMessage
+                self.write_value(value, characteristic.uuid)
+            except Exception as e:
+                logging.exception(f"Error executing callback: '{e}' - returning ''")
+                value = ""
+
+        logging.debug(
+            f"Read request for characteristc {characteristic}; returning '{value}'"
+        )
+
+        return to_byte_array(value)
+
+    def _write_request(
+        self, characteristic: BlessGATTCharacteristic, value: bytearray, **kwargs
+    ):
+        """method required by bless to handle write requests"""
+
+        logging.debug(
+            f"Write request for characteristic {characteristic.uuid} with value '{value}'"
+        )
+        # if handling a 'ChunkedMessage', callback is executed only when the message is complete
+        should_execute_callback = False
+
+        if ChunkedMessage.is_start_message(value):
+            self._partial_messages[characteristic.uuid] = ChunkedMessage()
+
+        if self._partial_messages.get(characteristic.uuid):
+            self._partial_messages[characteristic.uuid].append(value)
+        else:
+            should_execute_callback = True
+
+        # update characteristic value in all cases
+        self.write_value(value, characteristic.uuid)
```

Comment:
we definitely shouldn't need this one - you shouldn't be able to read/subscribe to a write characteristic in our model, and it might be slowing things down by writing all the incoming messages to the system bluetooth device. so just call the callback